lib/std/crypto/Certificate.zig

```diff
@@ -344,45 +344,88 @@ pub const Parsed = struct {
     // component or component fragment. E.g., *.a.com matches foo.a.com but
     // not bar.foo.a.com. f*.com matches foo.com but not bar.com.
     fn checkHostName(host_name: []const u8, dns_name: []const u8) bool {
+        // Empty strings should not match
+        if (host_name.len == 0 or dns_name.len == 0) return false;
+
+        // RFC 6125 Section 6.4.1: Exact match (case-insensitive)
         if (std.ascii.eqlIgnoreCase(dns_name, host_name)) {
             return true; // exact match
         }
 
-        var it_host = std.mem.splitScalar(u8, host_name, '.');
-        var it_dns = std.mem.splitScalar(u8, dns_name, '.');
+        // RFC 6125 Section 6.4.3: Wildcard certificates
+        // Wildcard must be leftmost label and in the form "*.rest.of.domain"
+        if (dns_name.len >= 3 and mem.startsWith(u8, dns_name, "*.")) {
+            const wildcard_suffix = dns_name[2..];
 
-        const len_match = while (true) {
-            const host = it_host.next();
-            const dns = it_dns.next();
+            // No additional wildcards allowed in the suffix
+            if (mem.indexOf(u8, wildcard_suffix, "*") != null) return false;
 
-            if (host == null or dns == null) {
-                break host == null and dns == null;
-            }
+            // Find the first dot in hostname to split first label from rest
+            const dot_pos = mem.indexOf(u8, host_name, ".") orelse return false;
+
+            // Wildcard matches exactly one label, so compare the rest
+            const host_suffix = host_name[dot_pos + 1 ..];
 
-            // If not a wildcard and they dont
-            // match then there is no match.
-            if (mem.eql(u8, dns.?, "*") == false and std.ascii.eqlIgnoreCase(dns.?, host.?) == false) {
+            // RFC 6125: Wildcards should not match IP addresses
+            // Use the standard library's IP parsing to check
+            if (std.net.Address.parseIp(host_name, 0)) |_| {
```

Comment:
Please do not add a dependency on std.net from within std.crypto.

RFC 6125 does not forbid matching IP addresses. It says that such things are "out of scope". Given that certificates are trusted, let's leave such decisions to certificate authorities and providers, and simply remove this check.